### ISSUE TITLE
Input data pipeline discussion

### ISSUE THREAD

**Issue by skydoorkai** [sql-machine-learning/elasticdl#276]
# An example of data preprocessing/augmentation for unstructured data training

Assume each data item is (jpg_image, label)  from recordio, and the input of the model is NHWC format of the images, with N = 16(batch_size), H=W=244, C=3 (RGB channel), and the preprocessing/augmentation pipeline is:

1. decode: jpg_image to float32(w, h, 3)

2. Augmentation: random flipping

3. Resize to (244, 244, 3)

4. Batching: accumulate 16 data samples to create a matrix with shape (16, 244, 244, 3), which will be the input of the model for training.

# Direct implementation
```
batch_size = 16
image_buf = []
label_buf = []

for record in rdio_r.get_reader(start, end):
    # need to provide an easy method for user to parse recordio item.
    jpg_image, label = parse(record)

    # --- Below should be provided by user for precessing/augmentation---#
    # 1. decode
    image = decode(jpg_image)
    # 2. random flipping for augmentation
    if random() < FLIP_PROBABILITY:
        image = horizontal flipping(image)
    # 3. resize
    image = resize(image, 244, 244)
    # --------------------------- end of user code-----------------------#

    # Batching
    image_buf.append(image)
    label_buf.append(label)
    if len(image_buf) == batch_size:
        images = tf.concat(image_buf, axis=0)
        images = tf.reshape(images, [batch_size, 244, 244, 3])
        labels = tf.concat(label_buf, axis=0)
        labels = tf.reshape(labels, [batch_size, 1])
        image_buf = []
        label_buf = []
        input_data = { 'image' : images, 'label' : labels}  # keys should be provided by user.
        # then start the minibatch training with input_data
        output = model.forward(input_data)
        loss = loss_func(output, input_data)
        ...
```

# Issues

- How to provide a user friendly way: user need to parse the recordio item, provide preprocessing/augmentation code, and the mapping (key-value) between input_data and the model's forward/loss function.  ElasticDL should implement the batching.

- Performance:  the items in a batch is processed one by one, which will be a bottleneck in training. 

- Also need to support prefetch: when training with one batch, the next batch's data can be processed simultaneously.
